README.md
#Deckard
Device OS and Browser detection.

[![Bower version](https://badge.fury.io/bo/deckard.svg)](http://badge.fury.io/bo/deckard)
[![Circle CI](https://circleci.com/gh/mobify/deckard.svg?style=shield&circle-token=4fa31dbc4e160e2b19aafb2d5881c617c3ea9b50)](https://circleci.com/gh/mobify/deckard)

![](http://media.tumblr.com/e0da98f48bf70afdc322f08794a70520/tumblr_inline_n20w2wbjJQ1qzrkdh.gif)

##Installation
Installation is simple via bower.

```
bower install https://github.com/mobify/deckard.git --save
```
##Configure with Require.js
Add the following configuration to your require.js config file.

```js
...
'deckard': 'bower_components/deckard/dist/deckard'
...
```
##Bring deckard in via Require.js

```js
define([ 
	'$', 
	'deckard'
], 
function($) {
	...
});
```

Bringing in `deckard` parses the User Agent string, and populates a number of properties related to the device. Additionally, `deckard` adds relevant classes to the HTML element, allowing you to target fixes via CSS.

Deckard runs automatically on your page if included (either via require.js or as a `<script>`).

###`$.os`

####Platform
type: **boolean**
- `$.os.desktop`
- `$.os.mobile`
- `$.os.tablet`

####OS Name
type: **boolean**
- `$.os.ios`
- `$.os.android`
- `$.os.windowsphone`
- `$.os.blackberry`
- `$.os.bb10`
- `$.os.rimtabletos`
- `$.os.kindle`

####OS Version
type: **number**
- `$.os.major`
- `$.os.minor`
- `$.os.patch`

type: **string**
- `$.os.version` (full version string)

###`$.retina`
type: **boolean**

###`$.browser`

####Browser Name
type: **boolean**
- `$.browser.safari`
- `$.browser.chrome`
- `$.browser.firefox`
- `$.browser.opera`
- `$.browser.ie`
- `$.browser.silk`
- `$.browser.webview`

####Browser Version
type: **number**
- `$.browser.major`
- `$.browser.minor`
- `$.browser.patch`

type: **string**
- `$.browser.version` (full version string)

###`$.orientation`
`deckard` also handles binding to `orientationchange`, and updates the CSS classes and properties
appropriately if that event is triggered.

type: **boolean**
- `$.orientation.landscape`
- `$.orientation.portrait`

You can then target a particular OS and Browser:

```js
if ($.os.android && $.os.major < 4 && $.browser.chrome) {
	// do stuff specific to android less than version 4 on chrome
}
```



CHANGELOG
1.0.2
 - Fix orientation detection in Chrome's mobile emulation mode
1.0.1
 - Removing unnecessary Velocity dependency in bower.json
1.0.0
 - Updating AMD wrapper
 - Refactor of methods inside deckard to decouple functionality


tests/runner/main.js

require(['config'], function() {
    require([
        'require',
        'chai',
        'mocha'
    ],
        function(require, chai, mocha) {

            var tests = [
                '../../tests/unit/deckardTests.js'
            ];

            require(tests, function() {
                assert = chai.assert;

                if (window.mochaPhantomJS) {
                    return window.mochaPhantomJS.run();
                }
                mocha.run();
            });
        });
});


tests/runner/config.js
require.config({
    baseUrl: '../../',
    paths: {
        'text': 'bower_components/requirejs-text/text',
        'fixtures': 'tests/fixtures',
        '$': 'lib/zeptojs/dist/zepto',
        'chai': 'node_modules/chai/chai',
        'mocha': 'node_modules/mocha/mocha'
    },
    'shim': {
        'mocha': {
            init: function() {
                this.mocha.setup('bdd');
                return this.mocha;
            }
        },
        '$': {
            exports: '$'
        }
    }
});


Gruntfile.js
'use strict'

var path = require('path');

module.exports = function(grunt) {
    var _ = grunt.util._;

    // By default, we load all local tasks from the tasks directory.
    grunt.file.expand('tasks/*').forEach(function(task) {
        grunt.loadTasks(task);
    });

    // Populate the config object
    var config = {};
    grunt.file.expand('tasks/config/*').forEach(function(configPath) {
        // Get the grunt-task name to put in the config which is based on the
        // name of the config file
        var configName = configPath.match(/\/([^\/]*)\.js/)[1];
        var option = require(path.join(__dirname + '/' + configPath))(grunt);
        config[configName] = _.extend(config[configName] || {}, option);
    });

    grunt.initConfig(_.extend({
        pkg: grunt.file.readJSON('package.json'),
        releaseName: '<%= pkg.name %>-<%= pkg.version %>',
        releaseMessage: '<%= pkg.name %> release <%= pkg.version %>'
    }, config));

    // load npm tasks
    var npmTasks = [
        'grunt-contrib-watch',
        'grunt-contrib-connect',
        'grunt-mocha-phantomjs',
        'grunt-contrib-copy',
        'grunt-contrib-uglify'
    ];

    npmTasks.forEach(function(taskName) {
        if (!grunt.task._tasks[taskName]) {
            grunt.loadNpmTasks(taskName);
            grunt.loadNpmTasks(taskName);
        }
    });

    grunt.registerTask('serve', ['build', 'connect:server', 'watch']);
    grunt.registerTask('build', ['lint:dev', 'copy', 'uglify']);
    grunt.registerTask('test', ['build', 'connect:test', 'mocha_phantomjs']);
    grunt.registerTask('test:browser', ['build', 'connect:test:keepalive']);
    grunt.registerTask('default', 'build');
};


src/js/detect.js
/**
 * Device OS and Browser detection
 *
 * Based on Zepto's detect.js
 * Zepto.js
 * (c) 2010-2014 Thomas Fuchs
 */
define([
    '$'
], function($) {
    var parseVersion = function(version) {
        if (!version) return {};

        var parts = version.split('.');

        return {
            version: version,
            major: parts[0] || 0,
            minor: parts[1] || 0,
            patch: parts[2] || 0
        };
    };

    var ua = window.navigator.userAgent;
    var $window = $(window);
    var $html = $('html');
    var orientationEvent = 'onorientationchange' in window ? 'orientationchange' : 'resize';
    // This ratio is less than 1 because it accommodates when keyboards are activated.
    var compareRatio = 0.8;

    /*jshint maxstatements:102 */
    var detect = function(ua) {
        var browserVersion;
        var osVersion;
        var os = this.os = {};
        var browser = this.browser = {};
        var orientation = this.orientation = {};
        var classes = [];
        var webkit = ua.match(/Web[kK]it[\/]{0,1}([\d.]+)/);
        var android = ua.match(/(Android);?[\s\/]+([\d.]+)?/);
        var osx = !!ua.match(/\(Macintosh\; Intel /);
        var ipad = ua.match(/(iPad).*OS\s([\d_]+)/);
        var ipod = ua.match(/(iPod)(.*OS\s([\d_]+))?/);
        var iphone = !ipad && ua.match(/(iPhone\sOS)\s([\d_]+)/);
        var windowsphone = ua.match(/Windows Phone ([\d.]+)/);
        var kindle = ua.match(/Kindle\/([\d.]+)/);
        var silk = ua.match(/Silk\/([\d._]+)/);
        var blackberry = ua.match(/(BlackBerry).*Version\/([\d.]+)/);
        var bb10 = ua.match(/(BB10).*Version\/([\d.]+)/);
        var rimtabletos = ua.match(/(RIM\sTablet\sOS)\s([\d.]+)/);
        var playbook = ua.match(/PlayBook/);
        var chrome = ua.match(/Chrome\/([\d.]+)/) || ua.match(/CriOS\/([\d.]+)/);
        var firefox = ua.match(/Firefox\/([\d.]+)/);
        var ie = ua.match(/MSIE\s([\d.]+)/) || ua.match(/Trident\/[\d](?=[^\?]+).*rv:([0-9.].)/);
        var webview = !chrome && ua.match(/(iPhone|iPod|iPad).*AppleWebKit(?!.*Safari)/);
        var safari = webview || ua.match(/Version\/([\d.]+)([^S](Safari)|[^M]*(Mobile)[^S]*(Safari))/);

        browser.webkit = !!webkit;

        if (browser.webkit) {
            browserVersion = webkit[1];
            classes.push('webkit');
        }

        if (android) {
            os.android = true;
            osVersion = android[2];
            classes.push('android');
        }
        if (iphone && !ipod) {
            os.ios = os.iphone = true;
            osVersion = iphone[2].replace(/_/g, '.');
            classes.push('ios iphone');
        }
        if (ipad) {
            os.ios = os.ipad = true;
            osVersion = ipad[2].replace(/_/g, '.');
            classes.push('ios ipad');
        }
        if (ipod) {
            os.ios = os.ipod = true;
            osVersion = ipod[3] ? ipod[3].replace(/_/g, '.') : null;
            classes.push('ios ipod');
        }
        if (windowsphone) {
            os.windowsphone = true;
            osVersion = windowsphone[1];
            classes.push('windows');
        }
        if (blackberry) {
            os.blackberry = true;
            osVersion = blackberry[2];
            classes.push('blackberry');
        }
        if (bb10) {
            os.bb10 = true;
            osVersion = bb10[2];
            classes.push('blackberry bb10');
        }
        if (rimtabletos) {
            os.rimtabletos = true;
            osVersion = rimtabletos[2];
            classes.push('blackberry');
        }
        if (playbook) {
            browser.playbook = true;
            classes.push('playbook');
        }
        if (kindle) {
            os.kindle = true;
            osVersion = kindle[1];
            classes.push('kindle');
        }
        if (silk) {
            browser.silk = true;
            browserVersion = silk[1];
            classes.push('silk');
        }
        if (!silk && os.android && ua.match(/Kindle Fire/)) {
            browser.silk = true;
            classes.push('silk');
        }
        if (chrome) {
            browser.chrome = true;
            browserVersion = chrome[1];
            classes.push('chrome');
        }
        if (firefox) {
            browser.firefox = true;
            browserVersion = firefox[1];
            classes.push('firefox');
        }
        if (ie) {
            browser.ie = true;
            browserVersion = ie[1];
            classes.push('ie');
        }
        if (safari && (osx || os.ios)) {
            browser.safari = true;
            classes.push('safari');
            if (osx) {
                browserVersion = safari[1];
            }
        }
        if (webview) {
            browser.webview = true;
            classes.push('webview');
        }

        os.tablet = !!(ipad || playbook || kindle || (android && !ua.match(/Mobile/)) ||
        (firefox && ua.match(/Tablet/)) || (ie && !ua.match(/Phone/) && ua.match(/Touch/)));

        os.mobile = !!(!os.tablet && !os.ipod && (android || iphone || blackberry || bb10 ||
        (chrome && ua.match(/Android/)) || (chrome && ua.match(/CriOS\/([\d.]+)/)) ||
        (firefox && ua.match(/Mobile/)) || (ie && ua.match(/Touch/))));

        // http://stackoverflow.com/questions/19689715/what-is-the-best-way-to-detect-retina-support-on-a-device-using-javascript
        os.retina = ((window.matchMedia && (window.matchMedia('only screen and (min-resolution: 192dpi), only screen and (min-resolution: 2dppx), only screen and (min-resolution: 75.6dpcm)').matches || window.matchMedia('only screen and (-webkit-min-device-pixel-ratio: 2), only screen and (-o-min-device-pixel-ratio: 2/1), only screen and (min--moz-device-pixel-ratio: 2), only screen and (min-device-pixel-ratio: 2)').matches)) || (window.devicePixelRatio && window.devicePixelRatio > 2)) && os.ios;
        os.retina && classes.push('retina');

        os = $.extend(true, os, parseVersion(osVersion));
        browser = $.extend(true, browser, parseVersion(browserVersion));

        classes.push(os.tablet ? 'tablet' : os.mobile ? 'mobile' : 'desktop');

        $html.addClass(classes.join(' '));
    };

    var orientation = function() {
        var isLandscape = ($window.height() / $window.width()) < compareRatio;

        $html.addClass('x-orientation-changing');

        if (isLandscape) {
            $html.removeClass('portrait').addClass('landscape');
            orientation.landscape = true;
            orientation.portrait = false;
        } else {
            $html.removeClass('landscape').addClass('portrait');
            orientation.landscape = false;
            orientation.portrait = true;
        }

        $html.removeClass('x-orientation-changing');
    };

    $window
        .on(orientationEvent, function() {
            orientation.call($);
        });

    detect.call($, ua);
    orientation.call($);

    $.__detect = detect;
    $.__orientation = orientation;
});


src/js/deckard.js
/**
 * Device OS and Browser detection
 *
 * Based on:
 *
 * Zepto detect.js
 * https://github.com/madrobby/zepto/blob/master/src/detect.js
 * (c) 2010-2014 Thomas Fuchs
 */
(function(factory) {
    if (typeof define === 'function' && define.amd) {
        define([
            '$'
        ], factory);
    } else {
        var framework = window.Zepto || window.jQuery;
        factory(framework);
    }
}(function($) {
    var _parseVersion = function(version) {
        if (!version) return {};

        var parts = version.split('.');

        return {
            version: version,
            major: parseInt(parts[0] || 0),
            minor: parseInt(parts[1] || 0),
            patch: parseInt(parts[2] || 0)
        };
    };

    var ua = window.navigator.userAgent;
    var device;
    var currentOrientation;
    var $window = $(window);
    var $html = $('html');
    var orientationEvent = 'onorientationchange' in window ? 'orientationchange' : 'resize';
    // This ratio is less than 1 because it accommodates when keyboards are activated.
    var compareRatio = 0.8;

    /*jshint maxstatements:120 */
    var detect = function(ua) {
        var browserVersion;
        var osVersion;
        var os = {};
        var browser = {};
        var cssClasses = [];

        var webkit = ua.match(/Web[kK]it[\/]{0,1}([\d.]+)/);
        var android = ua.match(/(Android);?[\s\/]+([\d.]+)?/);
        var osx = !!ua.match(/\(Macintosh\; Intel /);
        var ipad = ua.match(/(iPad).*OS\s([\d_]+)/);
        var ipod = ua.match(/(iPod)(.*OS\s([\d_]+))?/);
        var iphone = !ipad && ua.match(/(iPhone\sOS)\s([\d_]+)/);
        var windowsphone = ua.match(/Windows Phone ([\d.]+)/);
        var kindle = ua.match(/Kindle\/([\d.]+)/);
        var silk = ua.match(/Silk\/([\d._]+)/);
        var blackberry = ua.match(/(BlackBerry).*Version\/([\d.]+)/);
        var bb10 = ua.match(/(BB10).*Version\/([\d.]+)/);
        var rimtabletos = ua.match(/(RIM\sTablet\sOS)\s([\d.]+)/);
        var playbook = ua.match(/PlayBook/);
        var chrome = ua.match(/Chrome\/([\d.]+)/) || ua.match(/CriOS\/([\d.]+)/);
        var firefox = ua.match(/Firefox\/([\d.]+)/);
        var ie = ua.match(/MSIE\s([\d.]+)/) || ua.match(/Trident\/[\d](?=[^\?]+).*rv:([0-9.].)/);
        var webview = !chrome && ua.match(/(iPhone|iPod|iPad).*AppleWebKit(?!.*Safari)/);
        var safari = webview || ua.match(/Version\/([\d.]+)([^S](Safari)|[^M]*(Mobile)[^S]*(Safari))/);

        browser.webkit = !!webkit;

        if (browser.webkit) {
            browserVersion = webkit[1];
            cssClasses.push('webkit');
        }

        if (android) {
            os.android = true;
            osVersion = android[2];
            cssClasses.push('android');
        }
        if (iphone && !ipod) {
            os.ios = os.iphone = true;
            osVersion = iphone[2].replace(/_/g, '.');
            cssClasses.push('ios', 'iphone');
        }
        if (ipad) {
            os.ios = os.ipad = true;
            osVersion = ipad[2].replace(/_/g, '.');
            cssClasses.push('ios', 'ipad');
        }
        if (ipod) {
            os.ios = os.ipod = true;
            osVersion = ipod[3] ? ipod[3].replace(/_/g, '.') : null;
            cssClasses.push('ios', 'ipod');
        }
        if (windowsphone) {
            os.windowsphone = true;
            osVersion = windowsphone[1];
            cssClasses.push('windows');
        }
        if (blackberry) {
            os.blackberry = true;
            osVersion = blackberry[2];
            cssClasses.push('blackberry');
        }
        if (bb10) {
            os.bb10 = true;
            osVersion = bb10[2];
            cssClasses.push('blackberry', 'bb10');
        }
        if (rimtabletos) {
            os.rimtabletos = true;
            osVersion = rimtabletos[2];
            cssClasses.push('blackberry');
        }
        if (playbook) {
            browser.playbook = true;
            cssClasses.push('playbook');
        }
        if (kindle) {
            os.kindle = true;
            osVersion = kindle[1];
            cssClasses.push('kindle');
        }
        if (silk) {
            browser.silk = true;
            browserVersion = silk[1];
            cssClasses.push('silk');
        }
        if (!silk && os.android && ua.match(/Kindle Fire/)) {
            browser.silk = true;
            cssClasses.push('silk');
        }
        if (chrome) {
            browser.chrome = true;
            browserVersion = chrome[1];
            cssClasses.push('chrome');
        }
        if (firefox) {
            browser.firefox = true;
            browserVersion = firefox[1];
            cssClasses.push('firefox');
        }
        if (ie) {
            browser.ie = true;
            browserVersion = ie[1];
            cssClasses.push('ie');
        }
        if (safari && (osx || os.ios)) {
            browser.safari = true;
            cssClasses.push('safari');
            if (osx) {
                browserVersion = safari[1];
            }
        }

        if (webview) {
            browser.webview = true;
            cssClasses.push('webview');
        }

        os = $.extend(true, os, _parseVersion(osVersion));
        browser = $.extend(true, browser, _parseVersion(browserVersion));

        // Determines if this browser is the Android browser vs. chrome. It's always the
        // Android browser if it's webkit and the version is less than 537
        if (os.android && !browser.chrome && browser.webkit && browser.major < 537) {
            browser.androidBrowser = true;
            cssClasses.push('android-browser');
        }

        os.tablet = !!(ipad || playbook || kindle || (android && !ua.match(/Mobile/)) ||
        (firefox && ua.match(/Tablet/)) || (ie && !ua.match(/Phone/) && ua.match(/Touch/)));

        os.mobile = !!(!os.tablet && !os.ipod && (android || iphone || blackberry || bb10 ||
        (chrome && ua.match(/Android/)) || (chrome && ua.match(/CriOS\/([\d.]+)/)) ||
        (firefox && ua.match(/Mobile/)) || (ie && ua.match(/Touch/))));

        // http://stackoverflow.com/questions/19689715/what-is-the-best-way-to-detect-retina-support-on-a-device-using-javascript
        os.retina = ((window.matchMedia && (window.matchMedia('only screen and (min-resolution: 192dpi), only screen and (min-resolution: 2dppx), only screen and (min-resolution: 75.6dpcm)').matches || window.matchMedia('only screen and (-webkit-min-device-pixel-ratio: 2), only screen and (-o-min-device-pixel-ratio: 2/1), only screen and (min--moz-device-pixel-ratio: 2), only screen and (min-device-pixel-ratio: 2)').matches)) || (window.devicePixelRatio && window.devicePixelRatio > 2)) && os.ios;
        os.retina && cssClasses.push('retina');

        cssClasses.push(os.tablet ? 'tablet' : (os.mobile ? 'mobile' : 'desktop'));

        return {
            os: os,
            browser: browser,
            classes: cssClasses
        };
    };

    var orientation = function() {
        var isLandscape = ($window.height() / $window.width()) < compareRatio;

        return {
            orientation: {
                landscape: isLandscape,
                portrait: !isLandscape
            }
        };
    };

    var addClasses = function(device, orientation) {
        device.classes.push(orientation.landscape ? 'landscape' : 'portrait');
        $html.addClass(device.classes.join(' '));
    };

    $window
        .on(orientationEvent, function() {
            $.extend($, orientation());
            $html.toggleClass('portrait', $.orientation.portrait);
            $html.toggleClass('landscape', $.orientation.landscape);
        });

    device = detect(ua);
    currentOrientation = orientation();

    addClasses(device, currentOrientation.orientation);

    // bind os, browser, and orientation to $ as top-level properties
    $.extend($, { os: device.os, browser: device.browser }, currentOrientation);

    // expose for unit testing
    $.__deckard = {
        detect: detect,
        orientation: orientation,
        addClasses: addClasses
    };

    return $;
}));
